### PR TITLE
Added scripts for REACT_APP_BACKEND_WEBSOCKET_URL

### LINKED ISSUE #2183: Setup Script doesn't set the REACT_APP_BACKEND_WEBSOCKET_URL
**Describe the bug**

1. The setup script doesn't set the REACT_APP_BACKEND_WEBSOCKET_URL `.env` file variable
    1. This parameter is required for the proper functioning of plugins
1. The use of this parameter is not described in the `INSTALLATION.md` file

**To Reproduce**

1. Run the setup script. The parameter is not requested or updated
2. The parameter is not described in the `INSTALLATION.md` file

**Expected behavior**

1. The parameter must be automatically configured based on the value of `REACT_APP_TALAWA_URL` parameter value
2. The parameter must be described in the `INSTALLATION.md` file

**Actual behavior**

- See above

**Screenshots**

- N/A

**Additional details**

- N/A

**Potential internship candidates**

Please read this if you are planning to apply for a Palisadoes Foundation internship 
- https://github.com/PalisadoesFoundation/talawa/issues/359

### PR DESCRIPTION
**Pull request by adithyanotfound**
**Issue Number:**

Fixes #2183 

**Did you add tests for your changes?**

Yes.

**If relevant, did you update the documentation?**

Yes. (INSTALLATION.md)

**Summary**

1. Added scripts for REACT_APP_BACKEND_WEBSOCKET_URL.
2. Updated docs on how to setup the WebSocket URL.

**Does this PR introduce a breaking change?**

No.

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-admin/blob/master/CONTRIBUTING.md)?**

Yes.

## Summary by CodeRabbit

- **New Features**
	- Introduced functionality for configuring a WebSocket URL during the Talawa application setup.
	- Added instructions for setting the `REACT_APP_BACKEND_WEBSOCKET_URL` environment variable in the installation guide.

- **Documentation**
	- Enhanced `INSTALLATION.md` with detailed setup instructions for the WebSocket URL.
	- Updated table of contents and improved formatting for clarity and consistency.

- **Tests**
	- Added unit tests for the WebSocket URL configuration functionality to ensure proper conversion of HTTP/HTTPS URLs.

- **Chores**
	- Minor formatting adjustments made throughout the documentation for better readability.

### CODE REVIEW

File: setup.ts [hunk 1]
```diff
@@ -88,6 +89,45 @@ export async function main(): Promise<void> {
...
+    ).REACT_APP_BACKEND_WEBSOCKET_URL;
+    fs.readFile('.env', 'utf8', (err, data) => {
+      const result = data.replace(
+        `REACT_APP_BACKEND_WEBSOCKET_URL=${websocketUrl}`,
+        `REACT_APP_BACKEND_WEBSOCKET_URL=${endpoint}`,
+      );
+      fs.writeFileSync('.env', result, 'utf8');
+    });
+  }
+
```
[the remark this answers is not in the document]
Comment (adithyanotfound) [reply setup.ts:130]: Makes it less readable.

File: setup.ts [hunk 2]
```diff
@@ -88,6 +89,45 @@
...
   }
 
+  let shouldSetTalawaWebsocketUrl: boolean;
+
+  if (process.env.REACT_APP_BACKEND_WEBSOCKET_URL) {
+    console.log(
+      `\nTalawa WebSocket endpoint already exists: ${process.env.REACT_APP_BACKEND_WEBSOCKET_URL}`,
+    );
+    shouldSetTalawaWebsocketUrl = true;
+  } else {
```
[the remark this answers is not in the document]
Comment (adithyanotfound) [reply setup.ts:99]: Not required.

File: setup.ts [hunk 3]
```diff
@@ -88,6 +89,45 @@
...
+    const websocketUrl = dotenv.parse(
+      fs.readFileSync('.env'),
+    ).REACT_APP_BACKEND_WEBSOCKET_URL;
+    fs.readFile('.env', 'utf8', (err, data) => {
+      const result = data.replace(
+        `REACT_APP_BACKEND_WEBSOCKET_URL=${websocketUrl}`,
+        `REACT_APP_BACKEND_WEBSOCKET_URL=${endpoint}`,
+      );
+      fs.writeFileSync('.env', result, 'utf8');
+    });
```
[the remark this answers is not in the document]
Comment (adithyanotfound) [reply setup.ts]: Added error handling.

File: setup.ts [hunk 4]
```diff
@@ -6,6 +6,7 @@
 import { checkEnvFile } from './src/setup/checkEnvFile/checkEnvFile';
 import { validateRecaptcha } from './src/setup/validateRecaptcha/validateRecaptcha';
 import { askForCustomPort } from './src/setup/askForCustomPort/askForCustomPort';
+import { askForTalawaWebsocketUrl } from './src/setup/askForTalawaWebSocketUrl/askForTalawaWebSocketUrl';
```
[the remark this answers is not in the document]
Comment (adithyanotfound) [reply setup.ts]: Fixed.

**Review by palisadoes — CHANGES_REQUESTED** [review]
You have not followed the need in the issue. Please fix.

`The parameter must be automatically configured based on the value of REACT_APP_TALAWA_URL parameter value`

There should be no need to prompt for this parameter. Let me know if you can think of other scenarios where this may apply.

File: setup.ts [hunk 5]
```diff
@@ -86,6 +86,12 @@ export async function main(): Promise<void> {
       );
       fs.writeFileSync('.env', result, 'utf8');
     });
+
+    const websocketUrl = endpoint.replace(/^http(s)?:\/\//, 'ws$1://');
+    fs.appendFileSync(
+      '.env',
+      `REACT_APP_BACKEND_WEBSOCKET_URL=${websocketUrl}\n`,
+    );
```
[the remark this answers is not in the document]
Comment (adithyanotfound) [reply setup.ts]: Added that logic. Thanks

File: setup.ts [hunk 6]
```diff
@@ -74,18 +74,26 @@ export async function main(): Promise<void> {
...
+    const websocketUrl = endpoint.replace(/^http(s)?:\/\//, 'ws$1://');
+    const currentWebSocketUrl =
+      dotenv.parse(updatedEnvContent).REACT_APP_BACKEND_WEBSOCKET_URL;
+
+    const finalEnvContent = updatedEnvContent.replace(
+      `REACT_APP_BACKEND_WEBSOCKET_URL=${currentWebSocketUrl}`,
+      `REACT_APP_BACKEND_WEBSOCKET_URL=${websocketUrl}`,
+    );
+
+    fs.writeFileSync(envPath, finalEnvContent, 'utf8');
```
[the remark this answers is not in the document]
Comment (adithyanotfound) [reply setup.ts:96]: no required.

**Comment by adithyanotfound** [comment]
> You have not followed the need in the issue. Please fix.
> 
> `The parameter must be automatically configured based on the value of REACT_APP_TALAWA_URL parameter value`
> 
> There should be no need to prompt for this parameter. Let me know if you can think of other scenarios where this may apply.

I have made the required changes. Another possible scenario can be of the recaptcha key where it asks the user twice just to populate both `REACT_APP_USE_RECAPTCHA` and `REACT_APP_RECAPTCHA_SITE_KEY`.

Screenshot 2024-10-29 at 6 43 12 AM

**Review by palisadoes — CHANGES_REQUESTED** [review]
Thanks, but please update the tests for this.

**Comment by palisadoes** [comment]
Where are the test updates? The setup script has a test suite that you'll need to update

**Comment by adithyanotfound** [comment]
> Where are the test updates? The setup script has a test suite that you'll need to update

@palisadoes I'm working on it.

File: .env.example [hunk 7]
```diff
@@ -24,7 +24,7 @@ REACT_APP_USE_RECAPTCHA=
 REACT_APP_RECAPTCHA_SITE_KEY=
 
 # has to be inserted in the env file to use plugins and other websocket based features.
-REACT_APP_BACKEND_WEBSOCKET_URL=ws://localhost:4000/graphql
+REACT_APP_BACKEND_WEBSOCKET_URL=
```
Comment (palisadoes) [inline .env.example:27]: Why was this changed when it's a known working default value?

**Review by palisadoes — CHANGES_REQUESTED** [review]

File: .env.example (hunk 7 above)
Comment (adithyanotfound) [reply .env.example]: Changed it back to the default value.

**Comment by adithyanotfound** [comment]
@palisadoes Please review

**Pull request merged**